diff --git a/config/dwm/dwm.c b/config/dwm/dwm.c
index 73b3c1c..d7e333f 100644
--- a/config/dwm/dwm.c
+++ b/config/dwm/dwm.c
@@ -911,6 +911,8 @@ focus(Client *c)
 		XDeleteProperty(dpy, root, netatom[NetActiveWindow]);
 	}
 	selmon->sel = c;
+	if (selmon->lt[selmon->sellt]->arrange == monocle)
+		arrange(selmon);
 	drawbars();
 }
 
@@ -1223,11 +1225,14 @@ monocle(Monitor *m)
 			n++;
 	if (n > 0) /* override layout symbol */
 		snprintf(m->ltsymbol, sizeof m->ltsymbol, "[%d]", n);
-	for (c = nexttiled(m->clients); c; c = nexttiled(c->next))
-		if (selmon->pertag->drawwithgaps[selmon->pertag->curtag])
-			resize(c, m->wx, m->wy, m->ww - 2 * c->bw, m->wh - 2 * c->bw, 0);
-		else
-			resize(c, m->wx - c->bw, m->wy, m->ww, m->wh, False);
+
+	for (c = m->clients; c; c = c->next)
+		if (ISVISIBLE(c)) {
+			if (c == m->sel)
+				resize(c, m->wx, m->wy, m->ww - 2 * c->bw, m->wh - 2 * c->bw, 0);
+			else
+				XMoveWindow(dpy, c->win, WIDTH(c) * -2, m->wy); // move offscreen
+		}
 }
 
 void
